fabtests/hpcs/src/core/core.c

```diff
@@ -0,0 +1,1476 @@
+		if (order == CALLBACK_ORDER_EXPECTED && !state->all_recvs_done)
+			break;
+
+		prev = state->cur_receiver;
+		prev_threshold = state->cur_receiver_tx_threshold;
+
+		ret = pattern->next_receiver(arguments->pattern_arguments,
+				job->rank, job->ranks, &state->cur_receiver,
+				&state->cur_receiver_tx_threshold);
+		if (ret == -ENODATA) {
+			if (order == CALLBACK_ORDER_UNEXPECTED)
+				job->barrier();
+			state->all_sends_done = true;
+			break;
+		} else if (ret < 0) {
+			hpcs_error("next_receiver failed\n");
+			return ret;
+		}
+
+		if (state->tx_window == 0) {
+			state->cur_receiver = prev;
+			state->cur_receiver_tx_threshold = prev_threshold;
+			break;
+		}
+
+		op_context = CONTEXT(state->tx_context, state->sends_posted);
+		if (op_context->state != OP_DONE) {
+			state->cur_receiver = prev;
+			state->cur_receiver_tx_threshold = prev_threshold;
+			break;
+		}
+
+		test->tx_init_buffer(arguments->test_arguments,
+				DATA_BUF(state->tx_buf, state->sends_posted),
+				test_config->tx_buffer_size);
+
+		if (test_config->tx_use_cntr) {
+			mr = test->tx_create_mr(arguments->test_arguments,
+					domain_state->domain, 0,
+					DATA_BUF(state->tx_buf, state->sends_posted),
+					arguments->buffer_size, FI_SEND, 0);
+			if (mr == NULL) {
+				ret = -1;
+				hpcs_error("unable to register tx memory region\n");
+				return ret;
+			}
+
+			mr_desc = fi_mr_desc(mr);
+		} else {
+			mr = NULL;
+			mr_desc = NULL;
+		}
+
+		op_context->buf = DATA_BUF(state->tx_buf, state->sends_posted);
+		op_context->tx_mr = mr;
+
+		if (pattern->enable_triggered) {
+			for (i=0; i < test_config->tx_context_count; i++) {
+				op_context->ctxinfo[i].fi_trig_context.event_type = FI_TRIGGER_THRESHOLD;
+				op_context->ctxinfo[i].fi_trig_context.trigger.threshold =
+					(struct fi_trigger_threshold) {
+						.threshold =
+							state->recvs_done_prev +
+							(state->cur_receiver_tx_threshold * test_config->tx_context_count),
+						.cntr = domain_state->rx_cntr
+					};
+			}
+
+			op_context->tx_cntr = domain_state->tx_cntr;
+			op_context->domain = domain_state->domain;
+		}
+
+		ret = test->tx_transfer(arguments->test_arguments,
+				job->rank, 1,
+				domain_state->addresses[state->cur_receiver],
+				domain_state->endpoint, op_context,
+				op_context->buf, mr_desc, state->keys[state->cur_receiver],
+				job->rank, state->tx_flags);
+
+		if (ret == -FI_EAGAIN) {
+			state->cur_receiver = prev;
+			state->cur_receiver_tx_threshold = prev_threshold;
+			break;
+		}
+
+		hpcs_verbose("tx_transfer initiated from rank %ld "
+				"to rank %d: ctx %p key %ld trigger %d ret %d\n",
+				job->rank, state->cur_receiver, op_context,
+				state->keys[state->cur_receiver],
+				state->cur_receiver_tx_threshold, ret);
+
+		if (ret) {
+			hpcs_error("tx_transfer failed, ret=%d\n", ret);
+			return ret;
+		}
+
+		op_context->state = OP_PENDING;
+		op_context->core_context = state->sends_posted;
+
+		state->sends_posted++;
+		state->tx_window--;
+	};
+
+	return 0;
+}
```

Comment:
Above 2 functions are too long.  Limit functions to roughly one screen length and use sub-routines to improve readability.